File: servers/fai/src/fai/utils/jobs.py
import uuid
from collections.abc import (
    Awaitable,
    Callable,
)
from concurrent.futures import ThreadPoolExecutor
from datetime import (
    UTC,
    datetime,
)
from enum import Enum
from typing import Any

from sqlalchemy import (
    select,
    update,
)
from sqlalchemy.ext.asyncio import AsyncSession

from src.fai.db import async_session_maker
from src.fai.models.db.job_db import JobDb
from src.settings import LOGGER


class JobStatus(str, Enum):
    PENDING = "pending"
    IN_PROGRESS = "in_progress"
    COMPLETED = "completed"
    FAILED = "failed"


class JobManager:
    def __init__(self, max_workers: int = 4):
        self._executor = ThreadPoolExecutor(max_workers=max_workers)

    async def create_job(self, db: AsyncSession) -> str:
        job_id = str(uuid.uuid4())

        job = JobDb(id=job_id, status=JobStatus.PENDING, created_at=datetime.now(UTC))
        db.add(job)
        await db.commit()

        LOGGER.info(f"Created job {job_id}")
        return job_id

    async def execute_job(self, job_id: str, func: Callable[..., Awaitable[Any]], *args: Any, **kwargs: Any) -> None:
        async with async_session_maker() as db:
            await db.execute(
                update(JobDb)
                .where(JobDb.id == job_id)
                .values(status=JobStatus.IN_PROGRESS, started_at=datetime.now(UTC))
            )
            await db.commit()

            LOGGER.info(f"Starting job {job_id}")

            try:
                await func(*args, **kwargs)

                await db.execute(
                    update(JobDb)
                    .where(JobDb.id == job_id)
                    .values(status=JobStatus.COMPLETED, completed_at=datetime.now(UTC))
                )
                await db.commit()

                LOGGER.info(f"Job {job_id} completed successfully")

            except Exception as e:
                await db.execute(
                    update(JobDb)
                    .where(JobDb.id == job_id)
                    .values(status=JobStatus.FAILED, completed_at=datetime.now(UTC), error=str(e))
                )
                await db.commit()

                LOGGER.exception(f"Job {job_id} failed: {e}")

    async def get_job_status(self, db: AsyncSession, job_id: str) -> JobDb | None:
        result = await db.execute(select(JobDb).where(JobDb.id == job_id))
        return result.scalar_one_or_none()

    async def cleanup_old_jobs(self, db: AsyncSession, max_age_hours: int = 24) -> None:
        cutoff_timestamp = datetime.now(UTC).timestamp() - (max_age_hours * 3600)
        cutoff_datetime = datetime.fromtimestamp(cutoff_timestamp, tz=UTC)

        result = await db.execute(select(JobDb).where(JobDb.created_at < cutoff_datetime))
        old_jobs = result.scalars().all()
        LOGGER.info(f"Found {len(old_jobs)} old jobs to potentially clean up")


job_manager = JobManager()

File: packages/fern-docs/bundle/src/jobs.ts
import { NextResponse } from "next/server";

import { kv } from "@vercel/kv";

interface JobStatus {
  id: string;
  status: "pending" | "in_progress" | "completed" | "failed";
  created_at: number;
  started_at?: number;
  completed_at?: number;
  error?: string;
  result?: any;
}

export async function createJob(
  domain: string,
  messageId: string
): Promise<string> {
  const job: JobStatus = {
    id: messageId,
    status: "in_progress",
    created_at: Date.now(),
  };

  await kv.hset(domain, { [`tpuf_job`]: job });
  return job.id;
}

export async function getJobStatus(
  domain: string
): Promise<JobStatus | undefined> {
  const job = await kv.hget<JobStatus>(domain, `tpuf_job`);
  return job ?? undefined;
}

export function createJobResponse(
  job_id: string,
  status: "completed" | "failed" | "in_progress"
) {
  return NextResponse.json({
    job_id,
    status,
  });
}

export function createJobStatusResponse(
  jobId: string,
  status: "completed" | "failed" | "in_progress"
) {
  if (!status) {
    return NextResponse.json({ error: "Job not found" }, { status: 404 });
  }

  return NextResponse.json({
    id: jobId,
    status,
  });
}
